src/aspire/storage/starfile.py

```diff
@@ -1,173 +1,181 @@
-        return self.properties[name]
-
-    def __len__(self):
-        return len(self.loops)
-
-    def __getitem__(self, item):
-        return self.loops[item]
-
-    def __eq__(self, other):
-        return (
-            self.name == other.name
-            and self.properties == other.properties
-            and all([all(l1 == l2) for l1, l2 in zip(self.loops, other.loops)])
-        )
+class StarFileError(Exception):
+    pass
 
 
 class StarFile:
-    def __init__(self, starfile_path=None, blocks=None):
-
-        self.blocks = OrderedDict()
-
-        if starfile_path is not None:
-            self.init_from_starfile(starfile_path)
-        elif blocks is not None:
-            self.init_from_blocks(blocks)
-        else:
-            raise RuntimeError("Invalid constructor.")
-
-    def init_from_starfile(self, starfile_path):
+    def __init__(self, filepath="", blocks=None):
         """
-        Initalize a StarFile from a star file at a given path
-        :param starfile_path: Path to saved starfile.
-        :return: An initialized StarFile object
+        Initialize either from a path to a STAR file or from an OrderedDict of dataframes
         """
-        logger.info(f"Parsing starfile at path {starfile_path}")
-        with open(starfile_path, "r") as f:
-
-            blocks = []  # list of StarFileBlock objects
-            block_name = ""  # name of current block
-            properties = {}  # key value mappings to add to current block
-
-            loops = []  # a list of DataFrames
-            in_loop = False  # whether we're inside a loop
-            field_names = []  # current field names inside a loop
-            rows = []  # rows to add to current loop
-
-            for i, line in enumerate(f):
-                line = line.strip()
-
-                if line.startswith("#"):
-                    continue
-
-                # When in a 'loop', any blank line implies we break out of the loop
-                if not line:
-                    if in_loop:
-                        if rows:  # We have accumulated data for a loop
-                            loops.append(
-                                pd.DataFrame(rows, columns=field_names, dtype=str)
-                            )
-                            field_names = []
-                            rows = []
-                            in_loop = False
-
-                elif line.startswith("data_"):
-                    if loops or properties:
-                        blocks.append(
-                            StarFileBlock(loops, name=block_name, properties=properties)
+        self.blocks = blocks or OrderedDict()
+        self.filepath = str(filepath)
+        # the below syntax is an XOR of filepath and blocks. it will equal false if __init__
+        # is called with no arguments or with both kwargs and will equal true if either one or the other
+        # is passed
+        if not (bool(self.filepath) ^ bool(len(self.blocks))):
```

Comment:
Why not have default behavior (with no arguments) to just initialize with an empty `OrderedDict`?

Reply:
If the class doesn't have a way to incrementally add/extend then it doesn't make much sense to initialize it empty. The initial code Chris proposed, and this code, checked for that condition (empty file and no data).  I agree simplifying the logic would be nice.   Which do we think is the lesser evil here?

Reply:
Well if you initialize empty, `star = StarFile()`, you now have `star.blocks` as an empty OrderedDict. You could iteratively add blocks to it as needed via `star["new_blocks"] = xyz` where `xyz` is a DataFrame, and then write it to file once finished. Could be useful if you're not sure how many blocks / what the blocks will be yet.

Reply:
Okay yeah, if that all works then this makes sense to me now.  Thank you!